Framework/Algorithms/src/FindPeaksConvolve.cpp

```diff
@@ -0,0 +1,376 @@
+    if (iOverSigma(i) > m_iOverSigmaThreshold) {
+      closedData.emplace_back(i, iOverSigma(i));
+    } else if (iOverSigma(i) <= 0 || !m_performBinaryClosing) {
+      closedData.emplace_back(i, 0);
+    }
+  }
+  closedData.emplace_back(static_cast<int>(closedData.size()), 0); // Guarentee data region close
+  return closedData;
+}
+
+size_t FindPeaksConvolve::findPeakInRawData(const int xIndex, const TensorMap_const &yData,
+                                            size_t peakExtentBinNumber) {
+  peakExtentBinNumber = (peakExtentBinNumber % 2 == 0) ? peakExtentBinNumber + 1 : peakExtentBinNumber;
+  int sliceStart{xIndex - static_cast<int>(std::floor(static_cast<double>(peakExtentBinNumber) / 2.0))};
+  size_t adjPeakExtentBinNumber{peakExtentBinNumber};
+  int startAdj{0};
+
+  if (sliceStart < 0) {
+    startAdj = -sliceStart;
+    adjPeakExtentBinNumber = peakExtentBinNumber - startAdj;
+    sliceStart = 0;
+  }
+  if (sliceStart + adjPeakExtentBinNumber > static_cast<size_t>(yData.size()) - 1) {
+    adjPeakExtentBinNumber = static_cast<size_t>(yData.size()) - sliceStart;
+  }
+
+  Eigen::VectorXd::Index maxIndex;
+  if (m_findHighestDatapointInPeak) {
+    const auto unweightedYData = EigenMap_const(yData.data() + sliceStart, adjPeakExtentBinNumber);
+    unweightedYData.maxCoeff(&maxIndex);
+  } else {
+    Eigen::VectorXd pdf{generateNormalPDF(static_cast<int>(peakExtentBinNumber))};
+    const auto weightedYData = EigenMap_const(yData.data() + sliceStart, adjPeakExtentBinNumber)
+                                   .cwiseProduct(EigenMap_const(pdf.data() + startAdj, adjPeakExtentBinNumber));
+    weightedYData.maxCoeff(&maxIndex);
+  }
+  return static_cast<size_t>(maxIndex) + sliceStart;
+}
+
+Eigen::VectorXd FindPeaksConvolve::generateNormalPDF(const int peakExtentBinNumber) {
+  // assert vector has odd size.
+  Eigen::VectorXd pdf{peakExtentBinNumber};
+  boost::math::normal_distribution<> dist(0.0, peakExtentBinNumber / 2.0); // assures 2 stddevs in the resultant vector
+  const int meanIdx{peakExtentBinNumber / 2};
+  for (int i{0}; i < peakExtentBinNumber; ++i) {
+    int x{i - meanIdx};
+    pdf(i) = boost::math::pdf(dist, x);
+  }
+  return pdf;
+}
+
+void FindPeaksConvolve::createIntermediateWorkspaces(const size_t dataIndex, const Tensor1D &kernel,
+                                                     const Tensor1D &iOverSigma, const EigenMap_const &xData) {
+  API::Algorithm_sptr alg{createChildAlgorithm("CreateWorkspace")};
+
+  alg->setProperty("OutputWorkspace", "iOverSigma");
+  alg->setProperty("DataX", std::vector<double>(xData.data(), xData.data() + xData.rows()));
+  alg->setProperty("DataY", std::vector<double>(iOverSigma.data(), iOverSigma.data() + iOverSigma.size()));
+  alg->execute();
+  API::MatrixWorkspace_sptr algOutput = alg->getProperty("OutputWorkspace");
+  API::AnalysisDataService::Instance().add("iOverSigma" + std::to_string(dataIndex), algOutput);
```

Comment:
I've changed the two `add` calls in this function to `addOrRemove`. This stops runtime errors being thrown upon attempting to add duplicate workspaces.

Additionally, the workspace names are now <input workspace name>_iOverSigma_<index>.

I haven't pushed this change as I don't want to rerun tests, but will do when responding to review comments.